test/001_highsnr_stdio_audio/test_callback_rx.py

```diff
@@ -0,0 +1,186 @@
+import sys
+import argparse
+import numpy as np
+sys.path.insert(0,'../..')
+from tnc import codec2
+
+#--------------------------------------------GET PARAMETER INPUTS  
+parser = argparse.ArgumentParser(description='FreeDATA audio test')
+parser.add_argument('--bursts', dest="N_BURSTS", default=1, type=int)
+parser.add_argument('--framesperburst', dest="N_FRAMES_PER_BURST", default=1, type=int)
+parser.add_argument('--mode', dest="FREEDV_MODE", type=str, choices=['datac0', 'datac1', 'datac3'])
+parser.add_argument('--audiodev', dest="AUDIO_INPUT_DEVICE", default=-1, type=int,
+                    help="audio device number to use, use -2 to automatically select a loopback device")  
+parser.add_argument('--debug', dest="DEBUGGING_MODE", action="store_true")  
+parser.add_argument('--timeout', dest="TIMEOUT", default=10, type=int, help="Timeout (seconds) before test ends")  
+parser.add_argument('--list', dest="LIST", action="store_true", help="list audio devices by number and exit")  
+
+args = parser.parse_args()
+
+if args.LIST:
+    p = pyaudio.PyAudio()
+    for dev in range(0,p.get_device_count()):
+        print("audiodev: ", dev, p.get_device_info_by_index(dev)["name"])
+    quit()
+   
+N_BURSTS = args.N_BURSTS
+N_FRAMES_PER_BURST = args.N_FRAMES_PER_BURST
+AUDIO_INPUT_DEVICE = args.AUDIO_INPUT_DEVICE
+MODE = codec2.FREEDV_MODE[args.FREEDV_MODE].value
+DEBUGGING_MODE = args.DEBUGGING_MODE
+TIMEOUT = args.TIMEOUT
+
+# AUDIO PARAMETERS
+AUDIO_FRAMES_PER_BUFFER = 2400*2  # <- consider increasing if you get nread_exceptions > 0
+MODEM_SAMPLE_RATE = codec2.api.FREEDV_FS_8000
+AUDIO_SAMPLE_RATE_RX = 48000
+
+# make sure our resampler will work
+assert (AUDIO_SAMPLE_RATE_RX / MODEM_SAMPLE_RATE) == codec2.api.FDMDV_OS_48
+
+
+              
+# DATA CHANNEL INITIALISATION
+
+# open codec2 instance        
+freedv = cast(codec2.api.freedv_open(MODE), c_void_p)
+
+# get number of bytes per frame for mode
+bytes_per_frame = int(codec2.api.freedv_get_bits_per_modem_frame(freedv)/8)
+payload_bytes_per_frame = bytes_per_frame -2
+
+n_max_modem_samples = codec2.api.freedv_get_n_max_modem_samples(freedv)     
+bytes_out = create_string_buffer(bytes_per_frame * 2)
+
+codec2.api.freedv_set_frames_per_burst(freedv,N_FRAMES_PER_BURST)
+
+total_n_bytes = 0
+rx_total_frames = 0
+rx_frames = 0
+rx_bursts = 0
+rx_errors = 0
+nread_exceptions = 0
+timeout = time.time() + TIMEOUT
+receive = True
+audio_buffer = codec2.audio_buffer(AUDIO_FRAMES_PER_BUFFER*2)
+resampler = codec2.resampler()
+
+# Copy received 48 kHz to a file.  Listen to this file with:
+#   aplay -r 48000 -f S16_LE rx48_callback.raw
+# Corruption of this file is a good way to detect audio card issues
+frx = open("rx48_callback.raw", mode='wb')    
+
+# ------------------------------------------------ PYAUDIO CALLBACK
+def callback(data_in48k, frame_count, time_info, status):
+    global total_n_bytes
+    global rx_total_frames
+    global rx_frames
+    global rx_bursts
+    global receive
```

Comment:
:point_up: all these globals are messy, maybe put this all in class like `horus-gui`

Reply:
Class :ballot_box_with_check: